### DISCUSSION TITLE
[Proposal] / [Question] Why can't I directly cast a function pointer to void*?

### DISCUSSION THREAD

**Discussion by LostTime76** [dotnet/csharplang#7422] [Language Ideas]
I apologize if this question is dumb or has already been asked.. I have done a basic search, but I don't think I can find much.

I have a use case where I want to pass a bunch of managed function pointers into native code. I was wondering why I cannot directly cast the address of a static function into a void* or other long integer. Instead I have to cast to assign the explicit delegate first and then cast to a void*. This is not a show stopper, just an annoyance. I have to skirt the type checker? with an additional cast. Surely the type checker? knows that the result of the address of operation is a function pointer of given delegate type. Furthermore, I am announcing my explicit intent to type cast with the (void*) cast in unsafe code. When I have tens or hundreds of function pointers I want to pass to native code, duplicating a giant block of delegate casts instead of just being able to cast the static function to a void* directly is very cumbersome and ugly.

The ask would be to allow the direct cast into a void* or long integer in this case instead of having to go through the ceremony of the extra delegate cast when the type checker? I think doesn't need to do this. Is it because if the explicit delegate type is not announced, the & operator could produce one of several different delegate types (unmanaged / managed / cdecl, etc.)? Maybe we could have a default behavior in such cases. For example in this case, if you cast directly, the type checker assumes the same type as declared in this example (delegate* managed<void>).

![image](https://github.com/dotnet/csharplang/assets/62843610/8d331ca3-5041-4d76-a061-bda400724455)

**Comment by colejohnson66** [comment]
The [design document](https://github.com/dotnet/designs/blob/main/accepted/2022/FunctionPointers.md#using-void) for function pointers, unfortunately, does not mention _why_ this double cast is necessary

**Comment by 333fred** [comment]
Because natural function types (https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md#natural-function-type) only apply to a few specific scenarios. Address-of is not one of them.
